core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala

```diff
@@ -294,21 +295,20 @@ trait EntityComponent {
       }
     }
 
-    private def upsertAttributes(entities: Traversable[Entity], entityRecs: Traversable[EntityRecord], referencedAndSavingEntityRecs: Traversable[EntityRecord]) = {
-      val entityIds = entityRecs.map(_.id).toSeq
+    private def upsertAttributes(entitiesToSave: Traversable[Entity], entityIds: Seq[Long], entityIdsByRef: Map[AttributeEntityReference, Long]) = {
+      val attributesToSave = for {
+        entity <- entitiesToSave
+        (attributeName, attribute) <- entity.attributes
+        attributeRec <- entityAttributeQuery.marshalAttribute(entityIdsByRef(entity.toReference), attributeName, attribute, entityIdsByRef)
+      } yield attributeRec
 
-      def insertScratchAttributes(transactionId: String): ReadWriteAction[Unit] = {
-        val entityIdsByName = referencedAndSavingEntityRecs.map(r => AttributeEntityReference(r.entityType, r.name) -> r.id).toMap
-        val attributeRecsToEntityId = for {
-          entity <- entities
-          (attributeName, attribute) <- entity.attributes
-          attributeRec <- entityAttributeQuery.marshalAttribute(entityIdsByName(entity.toReference), attributeName, attribute, entityIdsByName)
-        } yield attributeRec
-
-        entityAttributeScratchQuery.batchInsertAttributes(attributeRecsToEntityId.toSeq, transactionId)
+      def insertScratchAttributes(attributeRecs: Seq[EntityAttributeRecord])(transactionId: String): ReadWriteAction[Unit] = {
+        entityAttributeScratchQuery.batchInsertAttributes(attributeRecs, transactionId)
       }
 
-      entityAttributeQuery.AlterAttributesUsingScratchTableQueries.upsertAction(entityIds, insertScratchAttributes)
+      entityAttributeQuery.findByOwnerQuery(entityIds) flatMap { existingAttributes =>
```

Comment:
this seems like a spot where repeatable read isolation level might be important. We have already read these attributes and are reading them again here. If they are different the second time the upsert will be incorrect. However, does our optimistic locking protect against this? The upsert may be wrong but the transaction will roll back. But then again, looking at our code, the first time we read the data (WorkspaceService.scala:653) we are not grabbing the entity version so the optimistic lock is not comprehensive enough.